smu/geometry/smu_molecule.py

```diff
@@ -0,0 +1,172 @@
+
+    # With the Hydrogens attached, the number of bonds to each atom.
+    self._bonds_with_hydrogens_attached = np.zeros((natoms), dtype=np.int32)
+    for bond in hydrogens_attached.bonds:
+      self._bonds_with_hydrogens_attached[bond.atom_a] += 1
+      self._bonds_with_hydrogens_attached[bond.atom_b] += 1
+
+    self._current_bonds_attached = np.zeros((natoms), dtype=np.int32)
+
+    # We turn bonds_to_scores into two arrays. So they can be iterated
+    # via itertools.
+
+    self._bonds = list(bonds_to_scores.keys())
+    self._scores = list(bonds_to_scores.values())
+
+    self._initial_score = 0.0
+    self._accumualate_score = operator.add
+
+    # For testing, it can be convenient to allow for partial matches
+    # For example this allows matching C-C and C=C without the need
+    # to add explicit hydrogens
+    self._must_match_all_bonds = matching_parameters.must_match_all_bonds
+
+  def set_initial_score_and_incrementer(self, initial_score: float, op: Callable) -> None:
+    """Update values used for computing scores"""
+    self._initial_score = initial_score
+    self._accumualate_score = op
+
+  def _initialize(self):
+    """Make the molecule reading for adding bonds between heavy atoms.
+    """
+    self._current_bonds_attached = np.copy(self._bonds_with_hydrogens_attached)
+
+  def _place_bond(self, a1: int, a2: int, btype: int) -> bool:
+    """Possibly add a new bond to the current config.
+
+    If the bond can be placed, updates self._current_bonds_attached for
+    both `a`` and `a2`.
+      Args:
+        a1:
+        a2:
+        btype:
+      Returns:
+    print(f"Trying to place bond {btype} current {self._current_bonds_attached[a1]} and {self._current_bonds_attached[a2]}")
+    """
+    if self._current_bonds_attached[a1] + btype > self._max_bonds[a1]:
+      return False
+    if self._current_bonds_attached[a2] + btype > self._max_bonds[a2]:
+      return False
+
+    self._current_bonds_attached[a1] += btype
+    self._current_bonds_attached[a2] += btype
+    return True
+
+  def generate_search_state(self) -> List[List[int]]:
+    """For each pair of atoms, return a list of plausible bond types.
+
+    This will be passed to itertools.product, which thereby enumerates all
+    possible bonding combinations.
+    Args:
+    Returns:
+      List of lists - one for each atom pair.
+    """
+    result: List[List[int]] = []
+    for ndx in range(0, len(self._bonds)):
+      # For each pair of atoms, the plausible bond types - non zero score.
+      plausible_types: List[int] = []
+      for i, score in enumerate(self._scores[ndx]):
+        if self._scores[ndx][i] > 0.0:
+          plausible_types.append(i)
+
+      result.append(plausible_types)
+
+    return result
+
+  def place_bonds(self, state: List[int]) -> Optional[dataset_pb2.BondTopology]:
+    """Place bonds corresponding to `state`.
+
+    Args:
+      state: for each pair of atoms, the kind of bond to be placed.
```

Comment:
If state is a List[int] how is it "for each pair of atoms"?

Reply:
Added a better explanation.